commit 75bbe6863f5f68567b2011fdbdee1cb6ddf19b79
Author: Kefu Chai <kchai@redhat.com>
Date:   Mon Feb 22 12:38:13 2021 +0800

    mgr: return the error if set_config() fails
    
    in general, `ActivePyModules::set_config()` is called by mgr module when
    serving user commands updating module, sometimes if the option is of the
    wrong type or invalid value, monitor rejects this request sent by mgr,
    but the error info is only logged in the logging message on mgr, but not
    returned to user. in this change, `ceph_set_module_option()` and the
    underlying methods are updated to return the error to the caller as an
    python exception.
    
    Signed-off-by: Kefu Chai <kchai@redhat.com>

diff --git a/src/mgr/ActivePyModules.cc b/src/mgr/ActivePyModules.cc
index e10546c2050..eb7c81b47e2 100644
--- a/src/mgr/ActivePyModules.cc
+++ b/src/mgr/ActivePyModules.cc
@@ -688,10 +688,12 @@ void ActivePyModules::set_store(const std::string &module_name,
   }
 }
 
-void ActivePyModules::set_config(const std::string &module_name,
-    const std::string &key, const boost::optional<std::string>& val)
+std::pair<int, std::string> ActivePyModules::set_config(
+  const std::string &module_name,
+  const std::string &key,
+  const boost::optional<std::string>& val)
 {
-  module_config.set_config(&monc, module_name, key, val);
+  return module_config.set_config(&monc, module_name, key, val);
 }
 
 std::map<std::string, std::string> ActivePyModules::get_services() const
diff --git a/src/mgr/ActivePyModules.h b/src/mgr/ActivePyModules.h
index f4feec4307a..d1e90f8bd3c 100644
--- a/src/mgr/ActivePyModules.h
+++ b/src/mgr/ActivePyModules.h
@@ -131,7 +131,7 @@ public:
 
   bool get_config(const std::string &module_name,
       const std::string &key, std::string *val) const;
-  void set_config(const std::string &module_name,
+  std::pair<int, std::string> set_config(const std::string &module_name,
       const std::string &key, const boost::optional<std::string> &val);
 
   PyObject *get_typed_config(const std::string &module_name,
